msgraph/client.go

```diff
@@ -78,17 +77,27 @@ type Client struct {
 	// This does not impact handling of retries related to rate limiting, which are always performed.
 	DisableRetries bool
 
-	httpClient *http.Client
+	// RequestMiddlewares is a slice of functions that are called in order before a request is sent
+	RequestMiddlewares *[]RequestMiddleware
+
+	// ResponseMiddlewares is a slice of functions that are called in order before a response is parsed and returned
+	ResponseMiddlewares *[]ResponseMiddleware
+
+	// HttpClient is the underlying http.Client, which by default uses a retryable client
+	HttpClient      *http.Client
+	retryableClient *retryablehttp.Client
 }
 
 // NewClient returns a new Client configured with the specified API version and tenant ID.
 func NewClient(apiVersion ApiVersion, tenantId string) Client {
+	r := retryablehttp.NewClient()
 	return Client{
-		Endpoint:   environments.MsGraphGlobal.Endpoint,
-		ApiVersion: apiVersion,
-		TenantId:   tenantId,
-		UserAgent:  "Hamilton (Go-http-client/1.1)",
-		httpClient: &http.Client{},
+		Endpoint:        environments.MsGraphGlobal.Endpoint,
+		ApiVersion:      apiVersion,
+		TenantId:        tenantId,
+		UserAgent:       "Hamilton (Go-http-client/1.1)",
+		HttpClient:      r.StandardClient(),
+		retryableClient: r,
 	}
 }
```

Comment:
unrelated to this PR, but a though that it'd be convenient to have a `NewClientWithEndpoint` constructor that this calls with `environments.MsGraphGlobal.Endpoint`?

Reply:
Nice, I like that. It has been asked for, but I was hesitant to overly clutter the `NewClient` function signature

Reply:
Although actually, to be used anywhere except corner cases, this would also have to be cascaded to every resource client